ssl/statem/extensions_clnt.c

```diff
@@ -1603,10 +1621,20 @@ int tls_parse_stoc_psk(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
         return 0;
     }
 
+    /*
+     * If we used the external PSK for sending early_data then s->early_secret
+     * is already set up, so don't overwrite it. Otherwise we copy the
+     * early_secret across that we generated earlier.
```

Comment:
As noted above you can send *both* a ticket based session and an external PSK based session in the ClientHello. Only one can be used for sending early_data, but either can be selected by the server and they will have different early secret values. This code deals with the case where we sent early_data using the ticket based session, but the server selected the external PSK instead. In that scenario the early secret that we used for sending the early_data is no longer valid - we need to use the one associated with the external PSK instead.